### ISSUE TITLE
Conversion between data.table (tabular) and array (n-dimensional) structural

### ISSUE THREAD

**Issue by myoung3** [Rdatatable/data.table#5061]
I often find myself needing to convert between data.table sand matrices or higher-dimensional arrays. In the case of 2-dimensional matrices, a `dcast` then followed by `as.matrix` works fine, but this doesn't generalize well to higher order data and even in the 2-dimensional case isn't the most efficient way to deal with this conversion.

It would be great if there were a convenience function to handle back and forth between these structures:

``` r
library(data.table)
set.seed(1)
dt <- CJ(dim1=1:3,dim2=1:4)
dt[, value:=rbinom(.N,10,prob=.5)]
dt
#>     dim1 dim2 value
#>  1:    1    1     4
#>  2:    1    2     4
#>  3:    1    3     5
#>  4:    1    4     7
#>  5:    2    1     4
#>  6:    2    2     7
#>  7:    2    3     7
#>  8:    2    4     6
#>  9:    3    1     6
#> 10:    3    2     3
#> 11:    3    3     4
#> 12:    3    4     4


structure(c(4L, 4L, 6L, 4L, 7L, 3L, 5L, 7L, 4L, 7L, 6L, 4L), .Dim = c(3L, 
4L, 1L), .Dimnames = list(c("1", "2", "3"), c("1", "2", "3", 
"4"), "value"))
#> , , value
#> 
#>   1 2 3 4
#> 1 4 4 5 7
#> 2 4 7 7 6
#> 3 6 3 4 4
```

**Comment by myoung3** [comment]
Here's some [preliminary work ](https://github.com/myoung3/fastreshape_matrix/blob/master/reshape_to_array.md) I've done on this problem.   Looking back on this I seemed particularly focused on the case of 3-dimensional data in a data.table where 2 dimensions are stored long and one dimension is stored wide. I don't remember my specific use-case for being interested in maximizing performance for that very specific structure, but I don't think it should be our focus. Rather, we should focus on the case where an arbitrary number of dimensions are all stored in the long tabular format in a data.table and converting that directly to an array.  The conversion here is fairly straight-forward since it's just keeping track of indices, but it would be nice to implement that once as part of the data.table package rather than having to figure this out every time I do it.

An important detail here: fwiu a vector and an array can't point to the same object in memory, despite the fact like it seems that should be possible (assuming the right dimension-major ordering of the vector). This means that the "value" column needs to be copied once in converting from data.table to array.

https://stackoverflow.com/questions/66021039/why-does-as-vector-deep-copy-a-matrix

**Comment by jangorecki** [comment]
We have array method, doesn't it work for your se case?

**Comment by myoung3** [comment]
regarding the existence of an array method, I can't seem to find it. What am I missing? This is version 1.14.0:
``` r
library(data.table)
set.seed(1)
dt <- CJ(dim1=1:3,dim2=1:4)
dt[, value:=rbinom(.N,10,prob=.5)]
array(dt) #returns a data.table??
#>          
#>  1: 1 1 4
#>  2: 1 2 4
#>  3: 1 3 5
#>  4: 1 4 7
#>  5: 2 1 4
#>  6: 2 2 7
#>  7: 2 3 7
#>  8: 2 4 6
#>  9: 3 1 6
#> 10: 3 2 3
#> 11: 3 3 4
#> 12: 3 4 4
class(array(dt))
#> [1] "data.table" "data.frame"

as.array(dt) #doesn't exist
#> Error in `dimnames<-.data.frame`(x, value): invalid 'dimnames' given for data frame
```

In any case, the following is the type of operation I think would be useful:
``` r
library(data.table)
set.seed(1)
dt <- CJ(dim1=1:3,dim2=1:4)
dt[, value:=rbinom(.N,10,prob=.5)]
dt
#>     dim1 dim2 value
#>  1:    1    1     4
#>  2:    1    2     4
#>  3:    1    3     5
#>  4:    1    4     7
#>  5:    2    1     4
#>  6:    2    2     7
#>  7:    2    3     7
#>  8:    2    4     6
#>  9:    3    1     6
#> 10:    3    2     3
#> 11:    3    3     4
#> 12:    3    4     4

setkey(dt, dim2,dim1)
stopifnot(sum(duplicated(dt[, dim1,dim2]))==0L)
unique_dim1 <- unique(dt[["dim1"]])
unique_dim2 <- unique(dt[["dim2"]])
array(dt$value,dim = c(length(unique_dim1),length(unique_dim2)),dimnames = list(unique_dim1, unique_dim2))
#>   1 2 3 4
#> 1 4 4 5 7
#> 2 4 7 7 6
#> 3 6 3 4 4
```

**Comment by jangorecki** [comment]
as.array

array function is not a method but a constructor.

**Comment by myoung3** [comment]
Did you see the error message above? There's no `data.table` method for `as.array`, at least in data.table 1.14.0. 
``` r
library(data.table)
set.seed(1)
dt <- CJ(dim1=1:3,dim2=1:4)
dt[, value:=rbinom(.N,10,prob=.5)]
as.array(dt)
#> Error in dim(x) <- length(x): invalid first argument
traceback()
#6: stop("invalid 'dimnames' given for data frame")
#5: `dimnames<-.data.frame`(x, value)
#4: `dimnames<-.data.table`(`*tmp*`, value = list(n))
#3: `dimnames<-`(`*tmp*`, value = list(n))
#2: as.array.default(dt)
#1: as.array(dt)
```

**Comment by myoung3** [comment]
There *does* seem to be an `as.data.table.array` that does basically the same thing as my `to_data.table` function above. 

So I guess I'm requesting the following:

- [ ] a help file  for `as.data.table.array` (so people knew what to expect in terms of output format and column names) 
- [ ] an `as.array.data.table` function (similar to my `to_array` function posted above) that can reverse the process of `as.data.table.array`

``` r
library(data.table)
a <- structure(c(4L, 4L, 6L, 4L, 7L, 3L, 5L, 7L, 4L, 7L, 6L, 4L), .Dim = c(3L, 
                                                                      4L, 1L), .Dimnames = list(c("1", "2", "3"), c("1", "2", "3", 
                                                                                                                    "4"), "1"))


as.data.table(a)
#>     V1 V2 V3 value
#>  1:  1  1  1     4
#>  2:  1  2  1     4
#>  3:  1  3  1     5
#>  4:  1  4  1     7
#>  5:  2  1  1     4
#>  6:  2  2  1     7
#>  7:  2  3  1     7
#>  8:  2  4  1     6
#>  9:  3  1  1     6
#> 10:  3  2  1     3
#> 11:  3  3  1     4
#> 12:  3  4  1     4
```

**Comment by jangorecki** [comment]
Thanks for checking that out. Agree it make sense.

**Comment by Mukulyadav2004** [comment]
Hi @MichaelChirico @jangorecki 
Not sure if a dedicated as.array.data.table is needed but documenting the existing array method behavior could still be useful for users to understand what to expect.
What do you think?

**Comment by jangorecki** [comment]
I think it can be useful. I have implemented it in
https://github.com/jangorecki/data.cube/blob/c8622b1c35d6d15557ca6fb99f9e597e728bf66c/R/data.table.R#L44
Could submit this to DT

**Comment by Mukulyadav2004** [comment]
Thanks for sharing the implementation! Would you recommend me to open a PR to submit this to DT. I'd be glad to assist if needed.

**Comment by jangorecki** [comment]
It needs wise detection of memory explosion. 10 rows 11 column data.table may be expanded to an array of size 8bytes * 10^10, filled in 99.9% with NAs

**Comment by Mukulyadav2004** [comment]
I thought of introducing a safety check that stops execution when memory requirements exceed a threshold and data sparsity falls below a minimum ratio. Do you think it is reasonable?

**Comment by jangorecki** [comment]
Yes, that's the way but what formula for the threshold you have in mind?

**Comment by Mukulyadav2004** [comment]
So by keeping threshold of memory as 2 GB and of data sparsity as 10% ,  we can formulate safety check as-
```
if(memory_gb > 2 && sparsity_ratio < 0.1){
  stop("memory explosion")
}
```

here `memory_gb` be the total memory required to store the complete array in `gb` which  we can calculate by multiplying the total number of array positions by 8 bytes and sparsity ratio is the ratio of actual data points compared to total possible array positions - we can calculate it by total no. of rows in dt by total combination of all unique dimension values.

e.g - if we have 10 data points but the array dimensions would create 10,000,000,000 possible positions, the sparsity ratio would be 10/10,000,000,000 = 0.000000001 which is much less than 10%.

**Comment by jangorecki** [comment]
Sounds like a good.plan. We could work on this branch together, I will bring data.cube implantation and you add threshold. Just give me couple days.

**Comment by ecoRoland2** [comment]
An arbitrary and fixed memory limit seems suboptimal. The user should at least have a possibility to change it.

**Comment by Mukulyadav2004** [comment]
You're right. To make it more user friendly, we can add parameters to as.array.data.table with these defaults that users can easily override by specifying their preferred values.
what do you think?

**Comment by jangorecki** [comment]
This could be even a global option, probably better than extra argument, which will be used in rare cases

**Comment by MichaelChirico** [comment]
For comparison, `CJ()` also has a tendency to produce row explosion, and it has no guardrails besides to ensure the output does not need long vector # of rows:

https://github.com/Rdatatable/data.table/blob/d8758232f92eb6f9ce6821a13c7ce0cdbac48680/R/setkey.R#L332

**Comment by Mukulyadav2004** [comment]
Its helpful. Should `as.array.data.table` follow this same pattern as CJ() - just checking integer overflow but not memory and sparsity?

**Comment by jangorecki** [comment]
> Its helpful. Should `as.array.data.table` follow this same pattern as CJ() - just checking integer overflow but not memory and sparsity?

IMO no.
CJ does not accept data.table, it is fot explicitly provided vectors. Passing full data.table object it is much more likely to get rows explosion.
